Implement Hecate

bs24 opened the pull request:
Principle used: let the opponent do a turn as if Hecate was not there. When confirmed, review the turn and stop it if it is illegal. Note that the opponent's endPlayerTurn() function has to be cancelled as well.

Some displays still look a bit weird but I don't know how to improve them.

Commit:
Nemesis

Commit:
ban Scylla vs Aphrodite

Commit:
ban ADONIS vs NEMESIS

Commit:
remove ban modification

Commit:
fix Nemesis vs Hydra + ban Nemesis vs Charybdis & Harpies as this 'force' implementation does not support opponent powers using a force during Nemesis's turn

quietmint committed:
Improved Nemesis, automatic second swap when possible

quietmint committed:
Reformat with spaces

quietmint committed:
Nemesis fix stats (each swap counts only once)

bs24 committed:
implement Tartarus

bs24 committed:
merge from main branch

bs24 committed:
merge from main branch

bs24 committed:
correct message when enters the abyss

bs24 committed:
minor

quietmint committed:
Improvements + make secret tokens generic

bs24 committed:
temp

bs24 committed:
Merge branch 'master' into hecate

quietmint committed:
Improve power getUiData (counter array)

quietmint committed:
Fix updateUI

quietmint committed:
Board automatically manages secret token height

bs24 committed:
temp before pull from main

bs24 committed:
Merge branch 'master' into hecate

bs24 committed:
Hecate should be mostly implemented but not tested

bs24 committed:
debug many Hecate matchups. Remains Medusa and argFirstPlayer

bs24 committed:
improve Hecate vs Medusa & Harpies

bs24 committed:
improve Hecate display

bs24 committed:
cp to original file name

bs24 committed:
remove aux files

bs24 committed:
merge

bs24 committed:
minor

bs24 committed:
correct Hecate vs Maenads & Aphrodite

quietmint commented:
> _When possible, use their power on their behalf to make their turns legal without informing them._

I wonder what does this mean in the description for Hecate? If the player moves or builds in an illegal space, we should not choose another space and move or build there instead. So if not about moving or building, I wonder in what case Hecate should use the player's power to change the turn?

> _and they lose the rest of their turn_

I assume this means any end-of-turn effect is cancelled as well?

quietmint commented:
I didn't fully review this code yet (hopefully I have more time tomorrow), but just thinking out loud, I wonder if it would be easier to check each action as it occurs (when the player moves/builds) and write a log entry like 'illegalAction' at that moment if the action conflicts with a secret worker. At turn end we look for the first 'illegalAction' to decide what steps to cancel.

bs24 commented:
> > _When possible, use their power on their behalf to make their turns legal without informing them._
> 
> I wonder what does this mean in the description for Hecate? If the player moves or builds in an illegal space, we should not choose another space and move or build there instead. So if not about moving or building, I wonder in what case Hecate should use the player's power to change the turn?
> 
> > _and they lose the rest of their turn_
> 
> I assume this means any end-of-turn effect is cancelled as well?

Hecate should never change the move or build space. Only "silent" powers can be used: Apollo and Minotaur can move into a space occupied by Hecate, by using their power. This means that Hecate, in this situation, must force her worker to the adequate space as if the power has been used. Iris (not implemented yet) can try to move further, and succeeds if there is a worker (Iris's or Hecate's) in the in-between space and none in the target space.

I do not think there is any other power concerned by this "use their power" rule. Charon could be one depending on the interpretation but it is banned. I also banned for the moment all powers targetting opponent workers  (e.g., Nemesis, Adonis, Odysseus, Siren) as the rule is not that clear.

And yes, the end-of-turn effect (mandatory or not) is also cancelled: e.g., Medusa, Hydra.

bs24 commented:
> I didn't fully review this code yet (hopefully I have more time tomorrow), but just thinking out loud, I wonder if it would be easier to check each action as it occurs (when the player moves/builds) and write a log entry like 'illegalAction' at that moment if the action conflicts with a secret worker. At turn end we look for the first 'illegalAction' to decide what steps to cancel.

I first thought about sth like this, which implies having afterOpponentMove, afterOpponentBuild, afterOpponentUsePower functions doing basically the same thing as the endOpponentTurn inner loop. I also do not know how deep in the log we should check the actions. As an illegal turn ends up with Hecate staying in the problematic space at the end of the turn for all powers so far, I though it would be easier to do it all at the end. I still had to use a tweak for Maenads & Medusa which would be treated more nicely in your proposition.

bs24 commented on santorini.game.php:
```diff
@@ -679,6 +689,10 @@ public function announceWin($playerId, $win = true)
+    // endgame: reveal all secret info 
+    $secrets = $this->board->getSecretPieces();
+    foreach ($secrets as $secret)
+      $this->notifyAllPlayers('revealPiece', '', ['piece' => $secret]); 
```
For Tartarus, this shows the abyss but does not update the UI. Not sure what the best way is to do it. Maybe move the pieces to the board then call the relevant power::updateUI()?

Also, what does the *Instant notification do? Is it more appropriate to call revealPieceInstant rather than revealPiece here?

quietmint replied:
Normal notification enforces a delay in the UI while the animation runs. It's maybe 1 or 2 seconds? In the main game .js file in setupNotifications function. The "instant" version doesn't delay the UI.

Three pieces in a loop without instant: revealed one by one, you can't see the second piece reveal until after the first piece completed the animation.

bs24 committed:
ban the bugged Hecate vs Chaos matchup

One way would be to block Chaos's endPlayerTurn when facing Hecate, then calling Chaos->endPlayerTurn inside Hecate's endOpponentTurn once the illegal actions have been cancelled.
Or create a inBetweenTurns functions for Chaos.

quietmint commented on modules/powers/Hecate.class.php:
```diff
@@ -13,10 +13,250 @@ public function __construct($game, $playerId)
+    // display current board state
+    $playerIds = $this->game->playerManager->getPlayerIds();
+    foreach ($playerIds as $playerId) {
+      $this->game->notifyPlayer($playerId, 'cancel', '' , [
```
FYI, this will break spectators. Since we do not know the playerId for spectators, the **only** way we communicate with them is via notifyAllPlayers. Probably this means we need to send two notifications -- one public notification to all players, and another private notification to each player with secret pieces.

quietmint commented on modules/powers/Hecate.class.php:
```diff
@@ -13,10 +13,250 @@ public function __construct($game, $playerId)
+    $this->game->notifyAllPlayers('revealPiece', '', ['piece' => $dummy]);
+    
+    // explain what is happening
+    $this->game->notifyAllPlayers('message', '${power_name}: due to a secret worker in ${coords}, an action of ${player_name} is illegal which cancels the rest of the turn', $args);
```
I think the display is improved if we show the message first (with the 'cancel' notification), so the user can read what is happening while the secret worker animation runs.

quietmint replied:
Also we can't forget `clienttranslate()` 🙂

quietmint committed:
Clean up endOpponentTurn and notification timing

quietmint commented:
So far, I reworked endOpponentTurn(). Mainly, I cleaned up notifications. I added 2 new parameters "duration" (so we no longer need "xxxInstant" notifications) and "ignorePlayerIds" (so we can send a notification to all players + spectators _except_ those in the list, no longer need dummy workers or special notifications "revealPiece", "hidePiece").

I did not review the move/build part of Hecate yet, but I think we should be able to greatly simplify this. Since Hecate move/build is the same as a player without any power, it would be much cleaner if we could just `return 'secret';` or somehow set `$args['location'] => 'secret';` so we could re-use the base code in `santorini.game.php`.

bs24 commented:
> no longer need dummy workers

Actually, I think we still need one so that we do not reveal the sex / orientation of the problematic secret worker. I had a comment about this but I guess I deleted it by mistake at some point. In the rules, you just know that there was a secret worker at the problematic space, but no other information on it. Putting a dummy one with a default orientation/sex  was one way to get away with it, hoping the opponent does not believe these are correct information about the actual worker.

I agree with your other comments, I did not know how to factorize the code more.

For Hecate move/build, we need to 1) look for secret workers, which is board->getPlacedWorkers($pId, true) 2) remove the other Hecate worker space as a possible work space 3) deal with the move notifications. It may indeed be easier to deal with it in the general code.

quietmint commented:
Okay in that case we can still use the real worker but change it's type to always female?

bs24 commented:
> Okay in that case we can still use the real worker but change it's type to always female?

Something like this would work as well, but we also need to hide the orientation somehow, as it reveals where the worker came from, right?

quietmint commented:
No, in board.js I think it uses a random direction for the place piece function?

Syarwin commented:
Yes,, it's random. That's also why there is a custom diff function when restarting your turn, because if you just erase and re-add, the orientation could change and that's a bit weird to see.

bs24 commented:
But to be completely secure (don't think it matters much TBH), we should not send the $piece if $piece['direction'] contains an information that the player should not know, even if it is not displayed by our js

quietmint commented:
Sure we can unset the direction

bs24 commented:
Actually, Syarwin told me the direction is not saved, so you can ignore all my comments about it! Making Hecate having two female workers or changing the revealed worker sex temporarily both work.

quietmint commented:
Got it. To preserve the power of Hippolyta, I will just do this for display purpose, not for the actual workers in hand.

bs24 commented:
do as you prefer, but just as a note, I don't think Hippolyta would be an issue here. She targets Hippolyta's female worker and not her opponent's (and Hecate is not a 4-player power). Actually, in the rules, Hecate has 2 identical worker tokens so there should be no rule issue.

quietmint commented:
The implementation with Minotaur is not correct. We are exposing the location of secret workers because we make those spaces unavailable for a normal move.

Hecate is standing at B2, C2:
![image](https://user-images.githubusercontent.com/622385/104077097-42cd2280-51e6-11eb-9b16-951a8828b6e9.png)

quietmint committed:
Hecate use common code for placeWorker

quietmint committed:
Hecate always display conflict worker as female

bs24 commented:
> The implementation with Minotaur is not correct. We are exposing the location of secret workers because we make those spaces unavailable for a normal move.

Ineed, I don't know how I missed that! I'll correct it today

Commit:
correct Minotaur: do not delete the normal argMove of both workers if one can force but not both

bs24 commented:
Corrected Minotaur, the bug happened if one Minotaur worker was secretly blocked by 2 Hecate workers and the other Minotaur worker could secretly do a force to the first Hecate worker.

bs24 commented:
Btw, do you think something should be displayed for Hecate when she sees an illegal display (two merged workers, a build on her worker...)? I don't know how, but it may be better to visually signal that the display is weird on purpose (e.g., highlight everything on this square in red)

quietmint commented:
I think no need. The opponent did actually move/build on that space, so having the messed-up display is kind of good. It clearly shows the action is not allowed. The display gets fixed once the turn is cancelled.

bs24 commented:
do you want to change the code further, or should I do a last round of tests with the problematic matchups?

quietmint committed:
32317: Fix restart turn for spectators

https://boardgamearena.com/bug?id=32317

quietmint commented:
Sorry, I have been busy with another game. Yes, let's test again and we can release if no issues.

bs24 commented:
I've seen a minor display issue when Hecate looses because of no legal moves: her worker are revealed so placed a second time for some reason then the display freezes. F5 solves it. At the end of any game she also sees her workers revealed.
Also, not sure if I did not update the files correctly for the tests but I need to refresh both pages after selecting the matchup. They seem to stay stuck in some loop.

Edit: except this and my last commit, seems good to me!

Commit:
fix Maenads vs Hecate + specify in the log that Hecate cancels end-of-turn actions. Not sure about this change though

quietmint commented:
I can't reproduce that problem you describe about double display for Hecate when no legal moves.

But for Hecate vs. Maenads, I think the current implementation is wrong.  Maenads should still win if she does an illegal build action (but not if she does an illegal move action). See here, Maenads moved B3 -> B4 to surround blue Hecate at C4, then did an illegal build on C3. The illegal build is cancelled, but she still surrounds Hecate, so she should still win.... Instead, her win is cancelled and the game continues. This cannot be correct...

![image](https://user-images.githubusercontent.com/622385/106226207-b2ff1080-61b4-11eb-8368-f10048af3752.png)

I think Hecate needs to check for powers with end-of-turn win conditions like this, and still trigger them again after cancelling the illegal actions. If the win condition is still true after cancelling the turn, it needs to happen....

quietmint commented:
Oh hmm, I just saw mlvanbie's comment https://discordapp.com/channels/523195652654497802/523214198680059926/793880436375158804 so I guess this behavior is correct, but it definitely feels wrong to me.

quietmint committed:
Hecate change the message
